### PR TITLE
Route TTS config through governed channel deploys

### PR DESCRIPTION
**Pull request by RasmusTho**
## Change Lane
- [ ] Docs authoring lane
- [ ] Governance lane

Final-Review-Rounds: 2

Governing-Issue: #4656

## Linked Issue
Closes #4656

## SBS Impact
- Primary subsystem: Product/Runtime release-channel deployment configuration
- Secondary subsystem(s): Builder System/CES deployment boundary and Companion TTS runtime
- Write class: current-state correction
- Persistence impact: No product persistence change; the existing untracked generated runtime environment is atomically replaced.
- Derived/rebuildable impact: No derived product artifact change; generated runtime configuration remains rebuildable from canonical machine-local inputs.
- New or changed contract: Channel deploys consume one redacted, fail-closed TTS selector snapshot before migration, pin, or Compose mutation; actuation refuses host-path creation and suppresses child output; rollback pins the disabled fallback independently of caller state.
- Owner-doc impact: Updated the TTS contract, provisioning runbook, deployment owner document, and operations owner surface in this PR.
- Transition debt impact: None; the prior caller-shell gap is removed and no compatibility shim is introduced.
- Boundary risk: High credential-durability and deployment-channel risk; production execution and host mutation are explicitly excluded.

## Owner-Doc Writeback
- [ ] No owner-doc change implied.
- [x] Owner-doc updated in this PR.
- [ ] Owner-doc follow-up issue created and linked.

## Summary
- Route machine-local TTS selectors through the channel-selected generated runtime environment and pin the validated snapshot into Compose without sourcing or CLI exposure.
- Atomically publish producer output, fail closed before channel mutation and after root-disappearance races, suppress governed Compose output, cover both producers and boundary states, and update the owning runtime and operator documentation.

## BuilderOps Routing
- Records/projections/receipts: none
- Reason: The live issue and this PR fully represent the bounded delivery; no divergence, learning, freshness, roadmap, or promotion material was created.

## Notes
Claim: task github-RasmusTho--agentic-pkm-mvp-issue-4656, lease lease-95f2c294, holder codex-sol-high. Validation: mechanism convergence CLEAN after the unchanged mechanism consumed 2 standard and 2 escalated repair attempts; 22 exact TTS config tests and 163 affected deploy/export/TTS tests passed; ruff, mypy (870 files), settings validation, docs guard, shell syntax, and diff checks passed. Prior-head review findings were repaired; two consecutive independent clean reviews are required on the current head. Config receipt is redacted to selector names, boolean/reason/path-class and fixed command results only. No deployment, production database, stable/prod ref movement, service restart, host mutation, or production secret/config change was performed.

### CODE REVIEW

File: scripts/lib/deploy_channel_compose.sh [hunk 1]
```diff
@@ -18,6 +18,189 @@ _deploy_channel_env_value() {
...
+try:
+    with os.scandir(candidate):
+        pass
+    resolved_candidate = candidate.resolve(strict=True)
+    resolved_root = root.resolve(strict=True)
+except OSError:
+    emit("blocked", "invalid_enabled_root", "inaccessible")
+    raise SystemExit(0)
+try:
+    candidate.relative_to(root)
```
Comment (RasmusTho) [inline scripts/lib/deploy_channel_compose.sh]: [P1] Remove the lexical containment rejection. The governing Issue requires the selected root to resolve outside the repository. An absolute spelling containing a parent component can be lexically relative to the repository while resolving to an existing sibling directory outside it, so this branch rejects contract-valid configuration. The resolved containment check below is the authority; add a positive regression case. This reclassifies discussion #3733588512 from P2 to protected P1 because it fails the governing acceptance contract.

File: docker-compose.yaml [hunk 2]
```diff
@@ -167,7 +167,8 @@ services:
       - "/Users:/Users"
       - "/Volumes:/Volumes"
       # External-SSD TTS store (models/cache/logs) bind-mounted to a fixed container path.
-      # Host root is machine-local (${TTS_HOST_ROOT} in .env.prod.local); defaults to a dev scratch dir.
+      # Host root is machine-local. Governed deploys pin it from the selected generated runtime env;
```
Comment (RasmusTho) [inline docker-compose.yaml:170]: [P1] Make the checked bind fail closed after preflight. The mount immediately below remains Compose short syntax, which may create a missing host source directory. If the validated directory disappears between preflight and actuation, deploy mutates host state and starts against a newly created empty root instead of failing. Use a long-form bind with `create_host_path: false` while preserving the disabled fallback, and add a disappearance-after-validation race regression.

File: scripts/lib/deploy_channel_compose.sh [hunk 3]
```diff
@@ -355,6 +528,14 @@ deploy_channel_compose() {
     # runtime env itself stays a service env_file; passing it as a CLI --env-file
     # would expose its DSNs and other values to Compose interpolation.
     export WATCHER_RUNTIME_ENV_FILE="${runtime_env_ref}"
+    if [ "${DEPLOY_TTS_CONFIG_GOVERNED:-0}" = "1" ]; then
```
Comment (RasmusTho) [inline scripts/lib/deploy_channel_compose.sh:527]: [P1] Adding the private selector to the Compose child environment requires an output-redaction boundary. The wrapper later streams that child stdout/stderr unchanged: a mount error can include the expanded private host path, and a `config` call through the helper renders resolved configuration including unrelated runtime values. The AC forbids raw paths, unrelated values/DSNs, and rendered Compose output on pass or failure. Capture/sanitize child diagnostics or reject output-producing modes, and test a failing child that emits the selector.

**Review by RasmusTho — COMMENTED** [review]
Independent final review round 1/2

Exact head: `854a43902a1397bd2771a0e57a8c7f3ee36508ec`

Verdict: BLOCKING

Severity summary: P0=0, P1=3, P2=0, P3=0.

The exact-head CI is green and the Bash 3.2 syntax check passes, but three protected contract failures remain: valid resolved-outside roots can be rejected, a checked bind can be created after the preflight race window, and Compose output is not redacted fail-closed. Under `verification-and-closure`, failed AC/config-durability/concurrency/redaction contracts are P1 and block. The earlier automated P2 suggestion on lexical containment is therefore classified here as P1, not an intended conservative boundary.

File: scripts/lib/deploy_channel_compose.sh (hunk 3 above)
Comment (RasmusTho) [reply scripts/lib/deploy_channel_compose.sh:527]: Fixed in  plus proof coverage in : governed Compose output is privately captured; rendered config is rejected, non-probe success output is suppressed, failures are fixed/redacted, and only validated internal container IDs may cross the boundary.

File: docker-compose.yaml (hunk 2 above)
Comment (RasmusTho) [reply docker-compose.yaml:170]: Fixed in  (current head ): the TTS mount now uses fail-closed long bind semantics with host-path creation disabled and a tracked disabled fallback. The disappearance race is covered and convergence is clean.

File: scripts/lib/deploy_channel_compose.sh [hunk 4]
```diff
@@ -18,6 +18,189 @@ _deploy_channel_env_value() {
...
+    resolved_root = root.resolve(strict=True)
+except OSError:
+    emit("blocked", "invalid_enabled_root", "inaccessible")
+    raise SystemExit(0)
+try:
+    candidate.relative_to(root)
+except ValueError:
+    pass
+else:
+    emit("blocked", "invalid_enabled_root", "repo_contained")
```
[the remark this answers is not in the document]
Comment (RasmusTho) [reply scripts/lib/deploy_channel_compose.sh]: Classified as protected P1 by independent final review and fixed on current head : resolved external selectors are accepted while resolved repository containment remains blocked. A positive regression now covers the normalized spelling.

File: scripts/lib/deploy_channel_compose.sh (hunk 1 above)
Comment (RasmusTho) [reply scripts/lib/deploy_channel_compose.sh]: Fixed in  (current head ): containment now follows the resolved candidate, with a positive normalized-external regression. The unchanged mechanism key passed the required convergence gate.

File: docker-compose.yaml (hunk 2 above)
Comment (RasmusTho) [reply docker-compose.yaml:170]: Correction: fixed in commit 9761160fe (current head 24ab8b6ff): the TTS mount now uses fail-closed long bind semantics with host-path creation disabled and a tracked disabled fallback. The disappearance race is covered and convergence is clean.

File: scripts/lib/deploy_channel_compose.sh (hunk 1 above)
Comment (RasmusTho) [reply scripts/lib/deploy_channel_compose.sh]: Correction: fixed in commit 9761160fe (current head 24ab8b6ff): containment now follows the resolved candidate, with a positive normalized-external regression. The unchanged mechanism key passed the required convergence gate.

File: scripts/lib/deploy_channel_compose.sh (hunk 3 above)
Comment (RasmusTho) [reply scripts/lib/deploy_channel_compose.sh:527]: Correction: fixed in commit 9761160fe plus proof coverage in 24ab8b6ff: governed Compose output is privately captured; rendered config is rejected, non-probe success output is suppressed, failures are fixed and redacted, and only validated internal container IDs may cross the boundary.

File: scripts/lib/deploy_channel_compose.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (RasmusTho) [reply scripts/lib/deploy_channel_compose.sh]: Correction: classified as protected P1 by independent final review and fixed on current head 24ab8b6ff. Resolved external selectors are accepted while resolved repository containment remains blocked; a positive regression covers the normalized spelling.

**Review by RasmusTho — COMMENTED** [review]
Independent final review clean round 1/2

Head: `24ab8b6ff8e051e1517216f488d442194e04f51e`

CLEAN

P0: 0
P1: 0
P2: 0
P3: 0

Current-head review covered all governing AC/Verify targets, atomic producer/snapshot behavior, shell error and cleanup paths, normalized/resolved containment, fail-closed bind behavior, output suppression/redaction and ID validation, all deploy/rollback consumers, secret bootstrap and Compose precedence, producer/fixture coverage, rollback/off behavior, and owner-doc writeback. Current CI is green. Focused configuration tests: 20 passed; shell syntax and diff checks passed.

File: docker-compose.yaml [hunk 5]
```diff
@@ -167,8 +167,15 @@ services:
...
-      - "${TTS_HOST_ROOT:-./tmp/tts}:/data/tts"
+      # Host root is machine-local. Governed deploys pin it from the selected generated runtime env;
+      # disabled/non-deploy compose consumers use the tracked empty fallback. Long syntax with
+      # create_host_path=false makes disappearance after preflight fail closed instead of creating
+      # a new empty host directory.
+      - type: bind
+        source: ${TTS_HOST_ROOT:-./config/tts-disabled}
+        target: /data/tts
+        bind:
+          create_host_path: false
```
Comment (RasmusTho) [inline docker-compose.yaml:178]: [P1] Preserve rollback availability from stale caller interpolation

The new long bind correctly refuses source creation, but rollback deliberately skips the TTS preflight and the Compose wrapper only pins or clears the TTS selectors when the governed-preflight flag is set. A stale caller `TTS_HOST_ROOT` that no longer exists therefore overrides the tracked disabled fallback and makes a real rollback `pull`/`up` fail before the prior service set can be recovered. That regresses the governing requirement that rollback remain unconditionally available; the current rollback test only uses `--dry-run`, so it never reaches Compose. Pin rollback to the disabled fallback (or otherwise clear the stale selector without invoking the deploy-only preflight) and add a hermetic non-dry-run rollback case.

**Review by RasmusTho — COMMENTED** [review]
Independent final review clean round 2/2

Exact head: `24ab8b6ff8e051e1517216f488d442194e04f51e`

Verdict: BLOCKING

Severity summary: P0=0, P1=1, P2=0, P3=0.

Current-head CI is green and the reviewed deploy path is otherwise fail-closed, but the actual rollback path can now be blocked by a stale caller selector because rollback bypasses the governed preflight while the new bind refuses a missing source. This fails the governing rollback-availability acceptance criterion and is protected P1.

File: docker-compose.yaml (hunk 5 above)
Comment (RasmusTho) [reply docker-compose.yaml:178]: Fixed on 2feb2d90889cb1b6188a7d056eb8d255a571a2c8. Non-dry rollback now pins governed TTS disabled state and clears any inherited host-root selector before every Compose path; a hermetic full rollback regression proves stale caller state cannot block rollback or create a host path. The follow-on convergence review also required and verified that an explicit blank selector fails redacted before any side effect. Full mechanism convergence is CLEAN after the fixed 2 standard plus 2 escalated attempt budget.

**Review by RasmusTho — COMMENTED** [review]
Independent final review round 1/2

Reviewed SHA: `2feb2d90889cb1b6188a7d056eb8d255a571a2c8`

Verdict: CLEAN

P0: 0
P1: 0
P2: 0
P3: 0

Missing tests: none against the governing Issue contract.

Residual risk: low. The deployment/configuration boundary remains fail-closed and was reviewed through current source, hermetic behavior tests, and green container CI; no live channel or production state was exercised.

Review covered every AC/Verify target, invariant-to-producer coverage, atomic snapshot publication and consumption, strict selector parsing, normalized/resolved containment, disappearance handling, Compose precedence, child-output redaction and validated probe IDs, secret-bootstrap wrapping, deploy and rollback state transitions, disabled/local-only fallback posture, owner-doc writeback, production exclusions, and proportional scope. Focused current-head verification passed together with shell syntax and diff checks; all current relevant CI is green.

**Review by RasmusTho — COMMENTED** [review]
Independent final review round 2/2

Reviewed SHA: `2feb2d90889cb1b6188a7d056eb8d255a571a2c8`

Verdict: CLEAN

P0: 0
P1: 0
P2: 0
P3: 0

Missing tests: none against the governing Issue contract.

Residual risk: low. The release/config boundary remains fail-closed and is covered by hermetic behavior tests and green container CI; no live channel, production database, host configuration, secret, ref, or service state was exercised.

Independent review covered every AC/Verify target; atomic snapshot publication and single-read consumption; read/decode/duplicate/strict-boolean failures; normalized external and resolved repository containment; inaccessible, missing, and disappearing roots; long-bind no-create behavior and disabled fallback; Compose environment precedence; command-output capture, fixed redaction, config rejection, and validated probe IDs; run/exec/up/pull/stop behavior with secret-bootstrap wrapping; deploy and rollback state transitions; invariant-to-producer and fixture coverage; owner-doc alignment; production exclusions; and proportional scope. Current-head focused validation passed 72 tests together with shell syntax and diff checks, and all current relevant CI is green.
